Support testing Processes with JUnit 5

ingorichtsmeier opened the pull request:
Provide an extension to run the ProcessEngine in Junit 5 tests and some sample tests to show how to use the extension.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda/camunda-bpm-platform?pullRequest=932) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/camunda/camunda-bpm-platform?pullRequest=932) before we can accept your contribution.<br/>**11** out of **16** committers have signed the CLA.<br/><br/>:white_check_mark: mboskamp<br/>:white_check_mark: tmetzke<br/>:white_check_mark: langfr<br/>:white_check_mark: fml2<br/>:white_check_mark: koevskinikola<br/>:white_check_mark: stefanruijsenaars<br/>:white_check_mark: bespaltovyj<br/>:white_check_mark: marfer<br/>:white_check_mark: vasicvuk<br/>:white_check_mark: jwulf<br/>:white_check_mark: ingorichtsmeier<br/>:x: tasso94<br/>:x: marstamm<br/>:x: ThorbenLindhauer<br/>:x: yanavasileva<br/>:x: camunda-jenkins<br/><hr/>**camunda-jenkins** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda/camunda-bpm-platform?pullRequest=932) it.</sub>

yanavasileva commented:
Hi @ingorichtsmeier,

Thank you for this PR.
We will have a look and get back to you.

Best regards,
Yana

yanavasileva commented:
Hello @ingorichtsmeier,

We will not be able to do a proper review now.
We would suggest providing JUnit5 support via community extension as the first step.
Then later (most probably during 7.15 development cycle) we could consider merging it to the engine.
Would you be interested in transforming the contribution to a community extension?

Best regards,
Yana

ingorichtsmeier commented:
Hi @yanavasileva,

have you decided this because of the large commit history?

In the  end I added just one class in the main folder `ProcessEngineExtension` (engine/src/main/java/org/camunda/bpm/engine/test/ProcessEngineExtension.java) and a method to the `org.camunda.bpm.engine.impl.test.TestHelper`:
```
public static String annotationDeploymentSetUp(ProcessEngine processEngine, Class<?> testClass, Method method, Deployment deploymentAnnotation, boolean onClass)
```
In the package engine/src/test/java/org/camunda/bpm/engine/test/standalone/testing I added 4 additional tests with the referenced process models. And the class ProvidedProcessEngineException under engine/src/test/java/org/camunda/bpm/engine/test/util/ProvidedProcessEngineExtension.java.

Maybe it's more effort to create a community extension then adding 10 classes on a clean branch to the engine code?

The benefit of adding them to the core is an easy extension in the camunda-bpm-process-test-coverage extension.

I can also do more refactoring on the TestHelper not to repeat the existing methods or move the code completly to the `ProcessEngineExtension`.

What do you think? 

I would like to make the review of the pull request as easy as possible for you.

Best regards, Ingo

ThorbenLindhauer commented:
Hi Ingo,

It is more about some design questions that we need to answer in the team before we can "just" review the code. For example:

* Do we want this to be in the core engine or should it be a separate module?
* Are the implemented features enough and are we also comfortable with supporting the feature set?

This makes the review here a bit larger and not something a developer can simply do while also working on other tasks. With respect to 7.14, we would therefore rather focus on our committed roadmap items for now. A community extension would allow you to make progress without waiting for us.

I hope that makes sense. It's also fine if you do not want to create a community extension. However, it may take us a while to review the contribution and we cannot guarantee that we can include it in 7.14

Best regards,
Thorben

jangalinski commented:
Great to see movement on  junit5 ! Tried this myself a while ago, but somehow lost focus.

I would vote for an additional library, instead of putting this in the engine ... I always thought that having my "compile"-scoped engine lib also providing the "test"-scoped engineRule was a violation of the SoC rule ... I would rather see junit4 moved out of the engine than junit5 moved into it ... 

Being author of several community extensions I also like to state: it's not that much effort and it's usually much quicker to extend and adopt than the "freighter" camunda-platform ... which is especially useful when the extension is fresh, features and versions change often and you do not want to wait for 6 month to get a release ... I would offer to help/support this as an extension ... which (once it's proven itself and becomes "feature-complete") could than move to the official support again (with all pros and cons).

tasso94 commented:
Hi @ingorichtsmeier,

I'm the community worker for this week, and I'm wondering how we proceed with this pull request. Are you planning to convert it into a community extension? If you think your contribution should go into the core engine, I would create a CAM issue and link it. 

Best,
Tassilo

ingorichtsmeier commented:
Hi @tasso94,

after @jangalinski s hint I prefer the community extension. 

I will move the code to a clean repository to start.

tasso94 commented:
Hi @ingorichtsmeier,

Alright! Happy to hear that. I'll close the pull request then. 🙂 

Best,
Tassilo
